enterprise/server/backends/authdb/authdb.go

```diff
@@ -25,19 +29,108 @@ import (
 
 const (
 	apiKeyLength = 20
+
+	// For encrypted keys, the first N characters serve as the nonce.
+	apiKeyNonceLength = 6
+
+	// Encrypted API keys have this prefix in the database.
+	apiKeyEncryptedValuePrefix = "en$"
+
+	apiKeyEncryptionBackfillBatchSize = 100
 )
 
 var (
 	userOwnedKeysEnabled = flag.Bool("app.user_owned_keys_enabled", false, "If true, enable user-owned API keys.")
+	apiKeyEncryptionKey  = flag.String("auth.api_key_encryption.key", "", "Base64-encoded encryption key for API keys.")
+	encryptNewKeys       = flag.Bool("auth.api_key_encryption.encrypt_new_keys", false, "If enabled, all new API keys will be written in an encrypted format.")
+	encryptOldKeys       = flag.Bool("auth.api_key_encryption.encrypt_old_keys", false, "If enabled, all existing unencrypted keys will be encrypted on startup.")
 )
 
 type AuthDB struct {
 	env environment.Env
 	h   interfaces.DBHandle
+
+	// Nil if API key encryption is not enabled.
+	apiKeyEncryptionKey []byte
+}
+
+func NewAuthDB(env environment.Env, h interfaces.DBHandle) (*AuthDB, error) {
+	adb := &AuthDB{env: env, h: h}
+	if *apiKeyEncryptionKey != "" {
+		key, err := base64.StdEncoding.DecodeString(*apiKeyEncryptionKey)
+		if err != nil {
+			return nil, status.InvalidArgumentErrorf("could not decode API Key encryption key: %s", err)
+		}
+		if len(key) != chacha20.KeySize {
+			return nil, status.InvalidArgumentError("API Key encryption key does not have expected length")
+		}
+		adb.apiKeyEncryptionKey = key
+		if err := adb.backfillUnencryptedKeys(); err != nil {
+			return nil, err
+		}
+	}
+	return adb, nil
 }
 
-func NewAuthDB(env environment.Env, h interfaces.DBHandle) *AuthDB {
-	return &AuthDB{env: env, h: h}
+func (d *AuthDB) backfillUnencryptedKeys() error {
+	if d.apiKeyEncryptionKey == nil {
+		return nil
+	}
+	if !*encryptOldKeys {
+		return nil
+	}
+
+	ctx := d.env.GetServerContext()
+	dbh := d.env.GetDBHandle().DB(ctx)
+
+	for {
+		query := fmt.Sprintf(`SELECT * FROM "APIKeys" WHERE value NOT LIKE '%s%%' LIMIT %d`, apiKeyEncryptedValuePrefix, apiKeyEncryptionBackfillBatchSize)
+		rows, err := dbh.Raw(query).Rows()
+		if err != nil {
+			return err
+		}
+		var keysToUpdate []*tables.APIKey
+		for rows.Next() {
+			var apk tables.APIKey
+			if err := dbh.ScanRows(rows, &apk); err != nil {
+				return err
+			}
+			keysToUpdate = append(keysToUpdate, &apk)
+		}
```

Comment:
uber nit: check `rows.Err()` after the `rows.Next()` loop. (IDK if there is ever an error in practice, but the `Next()` documentation says that `Err()` needs to be consulted to distinguish between there being no more rows left and `Next()` failing)